src/mappings/contracts.ts

```diff
@@ -246,16 +248,32 @@ export async function nodeContractCanceled(
 
   if (contractID === BigInt(0)) return
 
-  const savedContract = await ctx.store.get(NodeContract, { where: { contractID } })
+  const savedContract = await ctx.store.get(Contract, { where: { contractID } })
   if (!savedContract) return
 
   savedContract.state = ContractState.Deleted
-  await ctx.store.save<NodeContract>(savedContract)
+  await ctx.store.save<Contract>(savedContract)
+
+  let savedPublicIPs: PublicIp[] = await ctx.store.find(PublicIp, { where: { contractId: contractID }, relations: { farm: true } })
+  Promise.all(savedPublicIPs.map(async (ip) => {
+    ip.contractId = null
+    const farm = await ctx.store.get(Farm, { where: { farmID: ip.farm.farmID } })
+    if (farm) {
+      farm.freeIps += 1
```

Comment:
it's pretty common, i think, that a farms is filtered by how many public ips it has, and currently, farm queries in gridproxy counts all farm ips, and counts all farm ips that don't have a contract id (this is done per farm). and since the processor actually process's the events that manipulate ips, i thought it would be more efficient to have the ip counts pre-calculated.

i don't really see how this is error-prone, can you explain?